src/axom/core/detail/FlatTable.hpp

```diff
@@ -0,0 +1,333 @@
+      if(metadata.buckets[i] == GroupBucket::Empty)
+      {
+        return i;
+      }
+    }
+    // Bucket not found.
+    return InvalidSlot;
+  }
+
+  int nextFilledBucket(int start_index) const
+  {
+    for(int i = start_index + 1; i < Size; i++)
+    {
+      // We intentionally don't check for the sentinel here. This gives us the
+      // index of the sentinel bucket at the end, which is needed as a "stop"
+      // for forward iteration.
+      if(metadata.buckets[i] != GroupBucket::Empty)
+      {
+        return i;
+      }
+    }
+    return InvalidSlot;
+  }
+
+  template <typename Func>
+  int visitHashBucket(std::uint8_t hash, Func&& visitor) const
+  {
+    std::uint8_t reducedHash = reduceHash(hash);
+    for(int i = 0; i < Size; i++)
+    {
+      if(metadata.buckets[i] == reducedHash)
+      {
+        visitor(i);
+      }
+    }
+    return InvalidSlot;
+  }
+
+  void setBucket(int index, std::uint8_t hash)
+  {
+    metadata.buckets[index] = reduceHash(hash);
+  }
+
+  void clearBucket(int index) { metadata.buckets[index] = Empty; }
+
+  void setOverflow(std::uint8_t hash)
+  {
+    std::uint8_t hashOfwBit = 1 << (hash % 8);
+    metadata.ofw |= hashOfwBit;
+  }
+
+  bool getMaybeOverflowed(std::uint8_t hash) const
+  {
+    std::uint8_t hashOfwBit = 1 << (hash % 8);
+    return (metadata.ofw & hashOfwBit);
+  }
+
+  bool hasSentinel() const { return metadata.buckets[Size - 1] == Sentinel; }
+
+  void setSentinel() { metadata.buckets[Size - 1] = Sentinel; }
+
+  // We need to map hashes in the range [0, 255] to [2, 255], since 0 and 1
+  // are taken by the "empty" and "sentinel" values respectively.
+  static std::uint8_t reduceHash(std::uint8_t hash)
+  {
+    return (hash < 2) ? (hash + 8) : hash;
+  }
+
+  union alignas(16)
+  {
+    struct
+    {
+      std::uint8_t ofw;
+      std::uint8_t buckets[Size];
+    } metadata;
+    std::uint64_t data[2];
+    static_assert(sizeof(metadata) == sizeof(data),
+                  "flat_map::GroupBucket: sizeof(data_bytes) != sizeof(data)");
+  };
+};
+
+static_assert(sizeof(GroupBucket) == 16,
+              "flat_map::GroupBucket: size != 16 bytes");
+static_assert(std::alignment_of<GroupBucket>::value == 16,
+              "flat_map::GroupBucket: alignment != 16 bytes");
+static_assert(std::is_standard_layout<GroupBucket>::value,
+              "flat_map::GroupBucket: not standard layout");
+
+template <typename HashType, typename ProbePolicy = QuadraticProbing>
+struct SequentialLookupPolicy
+{
+  constexpr static int NO_MATCH = -1;
+
+  /*!
+   * \brief Inserts a hash into the first empty bucket in an array of groups
+   *  for an open-addressing hash map.
+   *
+   * \param [in] ngroups_pow_2 the number of groups, expressed as a power of 2
+   * \param [in] groups the array of metadata for the groups in the hash map
+   * \param [in] hash the hash to insert
+   */
+  IndexType probeEmptyIndex(int ngroups_pow_2,
+                            ArrayView<GroupBucket> groups,
```

Comment:
Suggestion: Would `metadata` be a better name instead of `groups`?  That's what how `FlatMap` refers to this parameter.  Same for the `probeIndex` method below.